online-shop/src/app/edit-product-form/edit-product-form.component.ts

```diff
@@ -0,0 +1,62 @@
+import { Location } from '@angular/common';
+import { Component, OnInit } from '@angular/core';
+import { FormBuilder, Validators } from '@angular/forms';
+import { ActivatedRoute } from '@angular/router';
+import { Product } from '../data/product';
+import { ProductService } from '../services/product.service';
+
+@Component({
+  selector: 'app-edit-product-form',
+  templateUrl: './edit-product-form.component.html',
+  styleUrls: ['./edit-product-form.component.scss'],
+})
+export class EditProductFormComponent implements OnInit {
+  productId = -1;
+  detailsForm = this.fb.nonNullable.group({
+    name: ['', Validators.required],
+    category: ['', Validators.required],
+    price: [0, [Validators.required, Validators.min(1)]],
+    image: ['', Validators.required],
+    description: ['', Validators.required],
+  });
+
+  constructor(
+    private fb: FormBuilder,
+    private route: ActivatedRoute,
+    private location: Location,
+    private productService: ProductService
+  ) {}
+
+  ngOnInit(): void {
+    this.productId = Number(this.route.snapshot.paramMap.get('id'));
+
+    this.productService
+      .getProduct(this.productId)
+      .subscribe((product) => this.detailsForm.patchValue(product));
+  }
+
+  applyEdits(): void {
+    const product: Product = {
+      id: this.productId,
+      name: this.detailsForm.value.name ?? '',
+      category: this.detailsForm.value.category ?? '',
+      price: Number(this.detailsForm.value.price ?? 0),
+      image: this.detailsForm.value.image ?? '',
+      description: this.detailsForm.value.description ?? '',
+    };
+
+    this.productService.updateProduct(product).subscribe({
+      next: () => {
+        alert('Product updated successfully!');
+        this.location.back();
+      },
+      error: () => {
+        alert('Failed to update product');
+      },
+    });
+  }
+
+  cancelEdits(): void {
+    this.location.back();
```

Comment:
**suggestion**: use angular built-in functionality by having `navigateByUrl` or `navigate` methods.